File: src/test/java/com/fwdekker/randomness/string/StringInsertActionTest.java
package com.fwdekker.randomness.string;

import com.fwdekker.randomness.CapitalizationMode;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.Arrays;
import java.util.Collection;
import java.util.HashSet;
import java.util.Set;
import java.util.regex.Pattern;

import static org.assertj.core.api.Assertions.assertThat;


/**
 * Parameterized unit tests for {@link StringInsertAction}.
 */
final class StringInsertActionTest {
    @SuppressWarnings("PMD.UnusedPrivateMethod") // Used as parameterized method source
    private static Collection<Object[]> provider() {
        return Arrays.asList(new Object[][]{
                {0, 0, "", CapitalizationMode.RETAIN, new Alphabet[]{}},
                {0, 0, "'", CapitalizationMode.UPPER, new Alphabet[]{}},
                {0, 0, "a", CapitalizationMode.LOWER, new Alphabet[]{}},
                {0, 0, "2Rv", CapitalizationMode.FIRST_LETTER, new Alphabet[]{}},

                {723, 723, "", CapitalizationMode.UPPER, new Alphabet[]{Alphabet.ALPHABET}},
                {466, 466, "z", CapitalizationMode.LOWER, new Alphabet[]{Alphabet.ALPHABET, Alphabet.UNDERSCORE}}
        });
    }


    @ParameterizedTest
    @MethodSource("provider")
    void testValue(final int minLength, final int maxLength, final String enclosure,
                   final CapitalizationMode capitalization, final Alphabet... alphabets) {
        final Set<Alphabet> alphabetSet = new HashSet<>(Arrays.asList(alphabets));

        final StringSettings stringSettings = new StringSettings();
        stringSettings.setMinLength(minLength);
        stringSettings.setMaxLength(maxLength);
        stringSettings.setEnclosure(enclosure);
        stringSettings.setCapitalization(capitalization);
        stringSettings.setAlphabets(alphabetSet);

        final StringInsertAction insertRandomString = new StringInsertAction(stringSettings);
        final Pattern expectedPattern
                = buildExpectedPattern(minLength, maxLength, enclosure, capitalization, alphabetSet);

        assertThat(insertRandomString.generateString()).containsPattern(expectedPattern);
    }


    private Pattern buildExpectedPattern(final int minLength, final int maxLength, final String enclosure,
                                         final CapitalizationMode capitalization, final Set<Alphabet> alphabets) {
        final StringBuilder regex = new StringBuilder();

        regex.append(enclosure);
        if (!alphabets.isEmpty()) {
            regex
                    .append('[')
                    .append(capitalization.getTransform().apply(Alphabet.concatenate(alphabets)))
                    .append("]{")
                    .append(minLength).append(',').append(maxLength)
                    .append('}');
        }
        regex.append(enclosure);

        return Pattern.compile(regex.toString());
    }
}

File: src/test/kotlin/com/fwdekker/randomness/string/StringInsertActionTest.kt
package com.fwdekker.randomness.string

import com.fwdekker.randomness.CapitalizationMode
import com.fwdekker.randomness.DataGenerationException
import com.fwdekker.randomness.DataGroupActionTest
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThat
import org.spekframework.spek2.Spek
import org.spekframework.spek2.style.specification.describe
import java.util.regex.Pattern


/**
 * Unit tests for [StringInsertAction].
 */
class StringInsertActionTest : Spek({
    describe("pattern") {
        data class Param(
            val minLength: Int,
            val maxLength: Int,
            val enclosure: String,
            val capitalization: CapitalizationMode,
            val symbolSets: Set<SymbolSet>
        )

        listOf(
            Param(0, 0, "", CapitalizationMode.RETAIN, setOf()),
            Param(0, 0, "'", CapitalizationMode.UPPER, setOf()),
            Param(0, 0, "a", CapitalizationMode.LOWER, setOf()),
            Param(0, 0, "2Rv", CapitalizationMode.FIRST_LETTER, setOf()),
            Param(723, 723, "", CapitalizationMode.UPPER, setOf(SymbolSet.ALPHABET)),
            Param(466, 466, "z", CapitalizationMode.LOWER, setOf(SymbolSet.ALPHABET, SymbolSet.UNDERSCORE))
        ).forEach { (minLength, maxLength, enclosure, capitalization, symbolSets) ->
            fun buildExpectedPattern(
                minLength: Int,
                maxLength: Int,
                enclosure: String,
                capitalization: CapitalizationMode,
                symbolSets: Set<SymbolSet>
            ): Pattern? {
                val base =
                    if (symbolSets.isEmpty()) ""
                    else "[${capitalization.transform.invoke(symbolSets.sum())}]{$minLength,$maxLength}"

                return base
                    .let { "$enclosure$it$enclosure" }
                    .let { Pattern.compile(it) }
            }

            it("generates a formatted string") {
                val stringScheme = StringScheme(
                    minLength = minLength,
                    maxLength = maxLength,
                    enclosure = enclosure,
                    capitalization = capitalization,
                    activeSymbolSets = symbolSets.toMap()
                )

                val insertRandomString = StringInsertAction(stringScheme)
                val expectedPattern = buildExpectedPattern(minLength, maxLength, enclosure, capitalization, symbolSets)

                assertThat(insertRandomString.generateString()).containsPattern(expectedPattern)
            }
        }

        it("retains emoji modifiers in the right order") {
            val emoji = "👩‍👩‍👧‍👧"
            val stringScheme = StringScheme(
                minLength = 1,
                maxLength = 1,
                enclosure = "",
                capitalization = CapitalizationMode.RETAIN,
                activeSymbolSets = setOf(SymbolSet("emoji", "👩‍👩‍👧‍👧")).toMap()
            )

            assertThat(StringInsertAction(stringScheme).generateString()).isEqualTo(emoji)
        }
    }

    describe("error handling") {
        it("throws an exception if the minimum is larger than the maximum") {
            val action = StringInsertAction(StringScheme(minLength = 99, maxLength = 21))
            Assertions.assertThatThrownBy { action.generateString() }
                .isInstanceOf(DataGenerationException::class.java)
                .hasMessage("Minimum length is larger than maximum length.")
        }

        it("throws an exception if no valid symbols are found") {
            val action = StringInsertAction(StringScheme(activeSymbolSets = emptyMap()))
            Assertions.assertThatThrownBy { action.generateString() }
                .isInstanceOf(DataGenerationException::class.java)
                .hasMessage("No valid symbols found in active symbol sets.")
        }
    }
})


/**
 * Unit tests for [StringGroupAction].
 */
class StringGroupActionTest : DataGroupActionTest({ StringGroupAction() })
